File: java/Strings/Rabin_Karp_Algorithm.java
package java.Strings;

public class Rabin_Karp_Algorithm {
  public int rabinKarp(String text, String pattern) {
    int n = text.length();
    int m = pattern.length();

    int d = 256;
    int q = 101;

    int h = 1;

    for (int i = 0; i < m - 1; i++) {
      h = (h * d) % q;
    }

    int p = 0;
    int t = 0;

    for (int i = 0; i <= n - m; i++) {
      p = (d * p + pattern.charAt(i)) % q;
      t = (d * t + text.charAt(i)) % q;
    }

    for (int i = 0; i <= n - m; i++) {
      if (p == t) {
        int j;
        for (j = 0; j < m; j++) {
          if (text.charAt(i + j) != pattern.charAt(j))
            break;
        }
        if (j == m)
          return i;
      }
      if (i < n - m) {
        t = (d * (t - text.charAt(i) * h) + text.charAt(i + m)) % q;
        if (t < 0)
          t += q;
      }
    }
    return -1;
  }
}

File: python/Strings/Rabin_Karp_Algorithm.py
def rabin_karp(text, pattern):
  d = 256
  q = 101 

  n = len(text)
  m = len(pattern)

  p = 0
  t = 0
  h = 1 

  for _ in range(m - 1):
    h = (h *d) % q 

  for i in range(m):
    p = (d * p + ord(pattern[i])) % q 
    t = (d * t + ord(text[i])) % q 

  for i in range(n - m + 1):
    if p == t:
      if text[i:i+m] == pattern:
        return i 

    if i < n - m:
      t = (d * (t - ord(text[i]) * h) + ord(text[i + m])) % q 

      if t < 0:
        t += q 
  
  return -1 
